client/state/preferences/reducer.js

```diff
@@ -0,0 +1,50 @@
+/**
+ * External dependencies
+ */
+import { combineReducers } from 'redux';
+import has from 'lodash/has';
+import get from 'lodash/get';
+import forOwn from 'lodash/forOwn';
+/**
+ * Internal dependencies
+ */
+import {
+	PREFERENCES_SET,
+	PREFERENCES_RECEIVE,
+	PREFERENCES_REMOVE,
+	PREFERENCES_FETCH
+} from 'state/action-types';
+import { USER_SETTING_KEY } from './actions';
+
+function values( state = {}, action ) {
+	if ( action.type === PREFERENCES_RECEIVE && has( action, [ 'data', USER_SETTING_KEY ] ) ) {
+		state = Object.assign( {}, state, get( action, [ 'data', USER_SETTING_KEY ] ) );
+		forOwn( state, ( value, key ) => {
+			if ( null === value ) {
+				delete state[ key ];
+			}
+		} );
+	} else if ( action.type === PREFERENCES_SET && action.key && action.value ) {
```

Comment:
Are we just trying to avoid using a switch here? :) I would probably stick with a switch until we choose a different reducer pattern. See #4805 and http://redux.js.org/docs/recipes/ReducingBoilerplate.html

Could we also move the validation inside of the if or move it to actions before we dispatch?